rust/src/delta.rs

```diff
@@ -540,14 +556,49 @@ impl DeltaTable {
             self.version(),
         );
 
-        while let PeekCommit::New(new_version, actions) =
-            self.peek_next_commit(self.version()).await?
+        // update to latest version if given max_version is not larger than current version
+        let max_version = max_version.filter(|x| x > &self.version());
+        let max_version: i64 = match max_version {
+            Some(x) => x,
+            None => self.get_latest_version().await?
+        };
+
+        let buf_size = self.config.log_buffer_size;
+
+        // construct stream yielding (version, bytes)
+        let store = self.storage.clone();
+        let log_stream = futures::stream::iter(self.version() + 1..).map(|i| {
+            let store2 = store.clone();
+            let loc = commit_uri_from_version(i).clone();
+            async move {
+                let out = store2.get(&loc);
+                (i, out.await)
+            }
+        });
+        
+        let mut log_buffer = {
+            let n_commits = usize::try_from(max_version - self.version());
+            match n_commits {
+                Ok(n) => log_stream.take(n).buffered(buf_size),
+                Err(err) => return Err(DeltaTableError::GenericError { source: Box::new(err) })
+            }
+        };
+        
+        while let Some((new_version, actions)) = {
+            let next_commit = log_buffer.next().await;
+            match next_commit {
+                Some((v, Ok(x))) => Ok(Some((v, self.get_actions(v,  x.bytes().await?).await?))),
+                Some((_, Err(ObjectStoreError::NotFound { .. }))) => Ok(None), // no more files in the log
+                Some((_, Err(err))) => Err(DeltaTableError::GenericError { source: Box::new(err) }),
+                None => Ok(None) // reached end of stream at max_version
+            }
+        }?
```

Comment:
It's a little hard to grok when there are so many lines on the RHS of a `while let`, that also ends with a `?`. Could we split out some of that logic above? Perhaps something like:

```suggestion
        let log_buffer = log_buffer
            .and_then(|(version, maybe_bytes)| async {
                let bytes = maybe_bytes?.await?;
                let actions = self.get_actions(version, bytes).await?;
                Ok((version, actions))
            });
        while let Some((new_version, actions)) = log_buffer.next().await? {
```

Also, shouldn't `NotFound` still be an error at this point since we previously verified the log file does exist?

Reply:
Yeah that's a good point about the while let. I think I tried to do this earlier, but for some reason I can't get the `and_then` method to work here even after importing `TryStreamExt`, I get complaints about not satisfying trait bounds.

The NotFound is an Ok case here if the max_version passed to the function is larger than the actual max_version, that's how it works currently on `main`. Not sure if that behavior is relied upon anywhere?

Reply:
Ah I forgot `max_version` is passed in. That part's fine then.

Reply:
This is a bit tricky, but this is what I'd propose (tested locally):

```rust
        let mut log_stream = futures::stream::iter(self.version() + 1..max_version + 1)
            .map(|version| {
                let store = store.clone();
                let loc = commit_uri_from_version(version);
                async move {
                    let data = store.get(&loc).await?.bytes().await?;
                    let actions = Self::get_actions(version, data).await?;
                    Ok((version, actions))
                }
            })
            .buffered(buf_size);

        while let Some(res) = log_stream.next().await {
            let (new_version, actions) = match res {
                Ok((version, actions)) => (version, actions),
                Err(DeltaTableError::ObjectStore {
                    source: ObjectStoreError::NotFound { .. },
                }) => break, // no more files in the log
                Err(err) => return Err(DeltaTableError::from(err)),
            };
```

As part of this, need to change `get_actions` to not take `&self`. I don't know why it does, given it does use `self` at all.

Reply:
Nice, updated the code to this. The `&self` was just an oversight from my part when factoring out the function.